### ISSUE TITLE
Keepalived 2.0.10 --- segfault

### ISSUE THREAD

**Issue by XFH1207** [acassen/keepalived#1706]
**Describe the bug**
VRRP_Script timed_out  occur segfault.

**Keepalived version**
Keepalived v2.0.10 (11/12,2018)

**Configuration file:**
vrrp_script check_ips {
    script "/usr/sbin/ping -w 5 172.18.0.41"
    timeout 2
    interval 5
}

vrrp_instance TBG01 {
    state BACKUP
    interface eth3
    virtual_router_id 51
    priority 100
    advert_int 1
    nopreempt

    virtual_ipaddress {
        172.17.1.1/32 dev eth0
        172.18.0.46/32 dev eth1
    }

    track_script {
        check_ips
    }
}

**System Log entries**
Jul 31 11:08:49 localhost Keepalived[5443]: Starting Keepalived v2.0.10 (11/12,2018)
Jul 31 11:08:49 localhost Keepalived[5443]: Running on Linux 4.18.0-147.el8.x86_64 #1 SMP Wed Dec 4 21:51:45 UTC 2019 (built for Linux 4.18.0)
Jul 31 11:08:49 localhost Keepalived[5443]: Command line: '/usr/sbin/keepalived' '-D'
Jul 31 11:08:49 localhost Keepalived[5443]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 31 11:08:49 localhost Keepalived[5444]: Starting Healthcheck child process, pid=5445
Jul 31 11:08:49 localhost Keepalived[5444]: Starting VRRP child process, pid=5446
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: Registering Kernel netlink reflector
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: Registering Kernel netlink command channel
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: Assigned address 10.10.197.1 for interface eth3
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: Registering gratuitous ARP shared channel
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: (TBG01) removing VIPs.
Jul 31 11:08:49 localhost Keepalived_healthcheckers[5445]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 31 11:08:49 localhost Keepalived_healthcheckers[5445]: Gained quorum 1+0=1 <= 1 for VS FWM 1011
Jul 31 11:08:49 localhost Keepalived_healthcheckers[5445]: Activating healthchecker for service [172.17.1.101]:none:80 for VS FWM 1011
Jul 31 11:08:49 localhost Keepalived_vrrp[5446]: VRRP sockpool: [ifindex(5), family(IPv4), proto(112), unicast(0), fd(12,13)]
Jul 31 11:08:51 localhost Keepalived_vrrp[5446]: **VRRP_Script(check_ips) timed_out**
Jul 31 11:08:51 localhost Keepalived_vrrp[5446]: (TBG01) Entering FAULT STATE
Jul 31 11:08:54 localhost Keepalived_healthcheckers[5445]: Misc check for [[172.17.1.101]:none:80 VS FWM 1011] by [/usr/bin/ping] failed (exited with status 1).
Jul 31 11:08:54 localhost Keepalived_healthcheckers[5445]: Disabling service [172.17.1.101]:none:80 to VS FWM 1011
Jul 31 11:08:54 localhost Keepalived_healthcheckers[5445]: Lost quorum 1-0=1 > 0 for VS FWM 1011
Jul 31 11:08:56 localhost kernel: keepalived[5446]: **segfault at 0 ip 000055d6130bd8bf sp 00007ffd1bd67c98 error 6 in keepalived[55d613060000+81000]**
Jul 31 11:08:57 localhost Keepalived[5444]: **Keepalived_vrrp exited due to segmentation fault (SIGSEGV).**
Jul 31 11:08:57 localhost Keepalived[5444]:  Please report a bug at https://github.com/acassen/keepalived/issues
Jul 31 11:08:57 localhost Keepalived[5444]:  and include this log from when keepalived started, a description
Jul 31 11:08:57 localhost Keepalived[5444]:  of what happened before the crash, your configuration file and the details below.
Jul 31 11:08:57 localhost Keepalived[5444]:  Also provide the output of keepalived -v, what Linux distro and version
Jul 31 11:08:57 localhost Keepalived[5444]:  you are running on, and whether keepalived is being run in a container or VM.
Jul 31 11:08:57 localhost Keepalived[5444]:  A failure to provide all this information may mean the crash cannot be investigated.
Jul 31 11:08:57 localhost Keepalived[5444]:  If you are able to provide a stack backtrace with gdb that would really help.
Jul 31 11:08:57 localhost Keepalived[5444]:  Source version 2.0.10
Jul 31 11:08:57 localhost Keepalived[5444]:  Built with kernel headers for Linux 4.18.0
Jul 31 11:08:57 localhost Keepalived[5444]:  Running on Linux 4.18.0-147.el8.x86_64 #1 SMP Wed Dec 4 21:51:45 UTC 2019
Jul 31 11:08:57 localhost Keepalived[5444]:  Command line: '/usr/sbin/keepalived' '-D'
Jul 31 11:08:57 localhost Keepalived[5444]:  configure options: --build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --program-prefix=
Jul 31 11:08:57 localhost Keepalived[5444]:                     --disable-dependency-tracking --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin
Jul 31 11:08:57 localhost Keepalived[5444]:                     --sbindir=/usr/sbin --sysconfdir=/etc --datadir=/usr/share
Jul 31 11:08:57 localhost Keepalived[5444]:                     --includedir=/usr/include --libdir=/usr/lib64 --libexecdir=/usr/libexec
Jul 31 11:08:57 localhost Keepalived[5444]:                     --localstatedir=/var --sharedstatedir=/var/lib --mandir=/usr/share/man
Jul 31 11:08:57 localhost Keepalived[5444]:                     --infodir=/usr/share/info --disable-libiptc --disable-ipset --enable-snmp
Jul 31 11:08:57 localhost Keepalived[5444]:                     --enable-snmp-rfc --enable-sha1 --with-init=systemd
Jul 31 11:08:57 localhost Keepalived[5444]:                     build_alias=x86_64-redhat-linux-gnu host_alias=x86_64-redhat-linux-gnu
Jul 31 11:08:57 localhost Keepalived[5444]:                     PKG_CONFIG_PATH=:/usr/lib64/pkgconfig:/usr/share/pkgconfig CFLAGS=-O2 -g -pipe
Jul 31 11:08:57 localhost Keepalived[5444]:                     -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS
Jul 31 11:08:57 localhost Keepalived[5444]:                     -fexceptions -fstack-protector-strong -grecord-gcc-switches
Jul 31 11:08:57 localhost Keepalived[5444]:                     -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1
Jul 31 11:08:57 localhost Keepalived[5444]:                     -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic
Jul 31 11:08:57 localhost Keepalived[5444]:                     -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection
Jul 31 11:08:57 localhost Keepalived[5444]:                     LDFLAGS=-Wl,-z,relro -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld
Jul 31 11:08:57 localhost Keepalived[5444]:  Config options: LVS VRRP VRRP_AUTH OLD_CHKSUM_COMPAT FIB_ROUTING SNMP_V3_FOR_V2 SNMP_VRRP SNMP_CHECKER
Jul 31 11:08:57 localhost Keepalived[5444]:                  SNMP_RFCV2 SNMP_RFCV3
Jul 31 11:08:57 localhost Keepalived[5444]:  System options: PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF LIBNL3 RTA_ENCAP
Jul 31 11:08:57 localhost Keepalived[5444]:                  RTA_EXPIRES RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID
Jul 31 11:08:57 localhost Keepalived[5444]:                  RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS FRA_L3MDEV FRA_UID_RANGE
Jul 31 11:08:57 localhost Keepalived[5444]:                  RTAX_FASTOPEN_NO_COOKIE RTA_VIA FRA_OIFNAME FRA_PROTOCOL FRA_IP_PROTO FRA_SPORT_RANGE
Jul 31 11:08:57 localhost Keepalived[5444]:                  FRA_DPORT_RANGE RTA_TTL_PROPAGATE IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS
Jul 31 11:08:57 localhost Keepalived[5444]:                  LWTUNNEL_ENCAP_ILA LIBIPTC_LINUX_NET_IF_H_COLLISION LIBIPVS_NETLINK
Jul 31 11:08:57 localhost Keepalived[5444]:                  IPVS_DEST_ATTR_ADDR_FAMILY IPVS_SYNCD_ATTRIBUTES IPVS_64BIT_STATS VRRP_VMAC
Jul 31 11:08:57 localhost Keepalived[5444]:                  SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SO_MARK SCHED_RT
Jul 31 11:08:57 localhost Keepalived[5444]:                  SCHED_RESET_ON_FORK
Jul 31 11:08:57 localhost Keepalived[5444]: VRRP child process(5446) died: Respawning
Jul 31 11:08:57 localhost Keepalived[5444]: Starting VRRP child process, pid=5453
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: Registering Kernel netlink reflector
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: Registering Kernel netlink command channel
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: Assigned address 10.10.197.1 for interface eth3
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: Registering gratuitous ARP shared channel
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: (TBG01) removing VIPs.
Jul 31 11:08:57 localhost Keepalived_vrrp[5453]: VRRP sockpool: [ifindex(5), family(IPv4), proto(112), unicast(0), fd(12,13)]
Jul 31 11:08:57 localhost systemd-coredump[5452]: Process 5446 (keepalived) of user 0 dumped core.#012#012Stack trace of thread 5446:#012#0  0x000055d6130bd8bf rb_erase (keepalived)#012#10x000055d6130b5ffd thread_child_handler (keepalived)#012#2  0x000055d6130bc6a8 signal_run_callback (keepalived)#012#3  0x000055d6130b7923 process_threads (keepalived)#012#4  0x000055d6130895cb start_vrrp_child (keepalived)#012#5  0x000055d61306babd keepalived_main (keepalived)#012#6  0x00007f916e52a873 __libc_start_main (libc.so.6)#012#7  0x000055d6130699fe _start (keepalive)
Jul 31 11:08:59 localhost Keepalived_vrrp[5453]: VRRP_Script(check_ips) timed_out
Jul 31 11:08:59 localhost Keepalived_vrrp[5453]: (TBG01) Entering FAULT STATE

**Additional context**
OS : CentOS Linux release 8.1.1911 (Core)
uname -a
Linux localhost.localdomain 4.18.0-147.el8.x86_64 #1 SMP Wed Dec 4 21:51:45 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Install mode**
yum install keepalived -y

**Comment by pqarmitage** [comment]
Can you please provide a stack backtrace from the coredump. `coredumpctl debug 5446` will invoke gdb on the coredump, and then `bt` will generate the backtrace. You will need to run `yum debuginfo-install keepalived` before using coredumpctl if you don't already have the keepalived-debuginfo and keepalived-debugsource packages installed.

**Comment by XFH1207** [comment]
I tried it again：
#### system log:
```
Jul 31 16:37:49 localhost Keepalived_healthcheckers[10191]: Misc check for [[172.17.1.101]:none:80 VS FWM 1011] by [/usr/bin/ping] failed (exited with status 1).
Jul 31 16:37:49 localhost Keepalived_healthcheckers[10191]: Disabling service [172.17.1.101]:none:80 to VS FWM 1011
Jul 31 16:37:49 localhost Keepalived_healthcheckers[10191]: Lost quorum 1-0=1 > 0 for VS FWM 1011
Jul 31 16:37:52 localhost kernel: **keepalived[10192]: segfault at 0 ip 0000557d9f44f8bf sp 00007fff95229ce8 error 6 in keepalived[557d9f3f2000+81000]**
Jul 31 16:37:52 localhost Keepalived[10190]: Keepalived_vrrp exited due to segmentation fault (SIGSEGV).
```
#### coredump:
```
# coredumpctl debug 10192
           PID: 10192 (keepalived)
           UID: 0 (root)
           GID: 0 (root)
        Signal: 11 (SEGV)
     Timestamp: Fri 2020-07-31 16:37:52 CST (27s ago)
  Command Line: /usr/sbin/keepalived -D
    Executable: /usr/sbin/keepalived
 Control Group: /system.slice/keepalived.service
          Unit: keepalived.service
         Slice: system.slice
       Boot ID: a469174adfa44930a69d3b0936181774
    Machine ID: 045c60765b1943b9b09d85bef83be060
      Hostname: localhost.localdomain
       Storage: /var/lib/systemd/coredump/core.keepalived.0.a469174adfa44930a69d3b0936181774.10192.1596184672000000.lz4
       Message: Process 10192 (keepalived) of user 0 dumped core.

                Stack trace of thread 10192:
                #0  0x0000557d9f44f8bf rb_erase (keepalived)
                #1  0x0000557d9f447ffd thread_child_handler (keepalived)
                #2  0x0000557d9f44e6a8 signal_run_callback (keepalived)
                #3  0x0000557d9f449923 process_threads (keepalived)
                #4  0x0000557d9f41b5cb start_vrrp_child (keepalived)
                #5  0x0000557d9f3fdabd keepalived_main (keepalived)
                #6  0x00007f702cb11873 __libc_start_main (libc.so.6)
                #7  0x0000557d9f3fb9fe _start (keepalived)

Failed to invoke gdb: No such file or directory
```

But I change config file: ping -w options
```
script "/usr/sbin/ping -w 5 172.18.0.41"
timeout 2 
interval 5
```
to:
```
script "/usr/sbin/ping -w 1 172.18.0.41"
timeout 2
interval 5
```
The segfault error did not occur. It may be related to a script execution timeout。

### Download keepalived-2.1.5.tar.gz
make done，use this version: 2.1.5
Modify ping -w options, the segfault error did not occur.
I'm sorry, due to the network, I do not provide keepalived-debuginfo installed, I think it might be the 2.0.10 version bugs.

**Comment by wangxp006** [comment]
Hi, according to the configuration provided by @XFH1207 before, this issue can be reproduced in the 2.0.10 . 
But using the 2.0.11, this problem cannot be reproduced. 
I believe this problem should have been fixed in later versions. 
The detailed call stack is as follows.

(gdb) bt
#0  rb_set_parent_color (rb=<optimized out>, color=1, p=0xaaadf81feb90) at rbtree_augmented.h:131
#1  ____rb_erase_color (augment_rotate=<optimized out>, root=<optimized out>, parent=0xaaadf81feb90) at rbtree.c:381
#2  rb_erase (node=node@entry=0xaaadf81e72a0, root=0xaaadf81e6c10) at rbtree.c:466
#3  0x0000aaadce7f3c1c in process_child_termination (status=15, pid=79271) at scheduler.c:1764
#4  thread_child_handler (v=<optimized out>, unused=<optimized out>) at scheduler.c:1794
#5  0x0000aaadce7f98f4 in signal_run_callback (thread=0xaaadf81e8ce0) at signals.c:343
#6  0x0000aaadce7f5264 in thread_call (thread=0xaaadf81e8ce0) at scheduler.c:1720
#7  process_threads (m=0xaaadf81e6ba0) at scheduler.c:1720
#8  0x0000aaadce7f5834 in launch_thread_scheduler (m=<optimized out>) at scheduler.c:1815
#9  0x0000aaadce7d0d68 in start_vrrp_child () at vrrp_daemon.c:1017
#10 start_vrrp_child () at vrrp_daemon.c:887
#11 0x0000aaadce7d0db0 in vrrp_respawn_thread (thread=<optimized out>) at vrrp_daemon.c:832
#12 0x0000aaadce7f5264 in thread_call (thread=0xaaadf81e7300) at scheduler.c:1720
#13 process_threads (m=0xaaadf81e7170) at scheduler.c:1720
#14 0x0000aaadce7f5834 in launch_thread_scheduler (m=<optimized out>) at scheduler.c:1815
#15 0x0000aaadce7b8114 in keepalived_main (argc=<optimized out>, argv=<optimized out>) at main.c:1897
#16 0x0000fffea20040fc in __libc_start_main () from /lib64/libc.so.6
#17 0x0000aaadce7b6228 in _start ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

**Comment by pqarmitage** [comment]
@XFH1207 and @wangxp006 Many thanks for the information you have both provided, along with the diagnosis that the problem is resolved in v2.0.11.

The issue was resolved by commit a577fa0c, so I suggest you either apply that patch to v2.0.10 or upgrade to a later version - v2.1.5 is the recommended version.

I will report this issue to RedHat so that they can incorporate the above patch.
